File: Electronic Store python/electronic_store.py
"""
Electronic Store - Python version
Converted from the Java "Electronic Store" project.

Python has real classes + inheritance, like Java, so the structure maps
closely:
    User (abstract base class, via the abc module) -> Customer, Owner,
    Staff, Supplier, Investor (all inherit from User)
    Product, Store, Review, Transaction -> plain classes, same as Java
"""

from abc import ABC, abstractmethod
from datetime import datetime


# ---------------------------- Review ----------------------------
class Review:
    counter = 1  # static counter, like Java's `static int counter`

    def __init__(self, rating, comments):
        self.id = Review.counter
        Review.counter += 1
        self.rating = rating
        self.comments = comments


# ---------------------------- Transaction ----------------------------
class Transaction:
    counter = 1  # static counter, like Java's `static int counter`

    def __init__(self, payment_type, amount):
        self.id = Transaction.counter
        Transaction.counter += 1
        self.payment_type = payment_type
        self.total_amount = amount
        self.date = str(datetime.now())


# ---------------------------- Product ----------------------------
class Product:
    def __init__(self, id, name, price, warranty_period):
        self.id = id
        self.name = name
        self.price = price
        self.discount = 0
        self.warranty_period = warranty_period
        self.reviews = []

    # encapsulation is used for the detail of the product data
    def get_name(self):
        return self.name

    def get_price(self):
        return self.price - self.discount

    def set_discount(self, discount):
        self.discount = discount

    def add_review(self, review):
        self.reviews.append(review)


# ---------------------------- Store ----------------------------
class Store:
    def __init__(self, id, name, location):
        self.id = id
        self.name = name
        self.location = location
        self.products = []

    def get_name(self):
        return self.name

    def stock_product(self, product):
        self.products.append(product)
        print(f"{product.get_name()} stocked in {self.name}")


# ---------------------------- User (abstract base class) ----------------------------
# so this is the parent class and it is an abstract class because it has an
# abstract method display_role() which will be implemented by the child classes
class User(ABC):
    def __init__(self, id, name, email, phone, age, address):
        self.id = id
        self.name = name
        self.email = email
        self.phone = phone
        self.age = age
        self.address = address

    # encapsulation is used for the detail of the users data
    def get_id(self):
        return self.id

    def get_name(self):
        return self.name

    def set_name(self, name):
        self.name = name

    def get_email(self):
        return self.email

    def set_email(self, email):
        self.email = email

    # Abstraction
    @abstractmethod
    def display_role(self):
        pass


# ---------------------------- Customer ----------------------------
class Customer(User):
    def __init__(self, id, name, email, phone, age, address):
        super().__init__(id, name, email, phone, age, address)
        self.transactions = []
        self.reviews = []

    def display_role(self):
        print("Role: Customer")

    def buy_product(self, product, store, payment_type):
        t = Transaction(payment_type, product.get_price())
        self.transactions.append(t)
        print(f"{self.get_name()} bought {product.get_name()} from {store.get_name()}")

    def write_review(self, product, rating, comment):
        r = Review(rating, comment)
        self.reviews.append(r)
        product.add_review(r)
        print(f"{self.get_name()} wrote a review for {product.get_name()}")


# ---------------------------- Owner ----------------------------
# so here we have created a child class of the User class
class Owner(User):
    def __init__(self, id, name, email, phone, age, address, equity):
        super().__init__(id, name, email, phone, age, address)
        self.equity = equity
        self.managed_stores = []

    # Implement the abstract method from the User class
    def display_role(self):
        print(f"Role: Owner with {self.equity} equity")

    # Method to add a store to the list of managed stores
    def add_store(self, store):
        self.managed_stores.append(store)
        print(f"{self.get_name()} now manages {store.get_name()}")


# ---------------------------- Supplier ----------------------------
class Supplier(User):
    def __init__(self, id, name, email, phone, age, address, company_name, price_value):
        super().__init__(id, name, email, phone, age, address)
        self.company_name = company_name
        self.supplied_products = []
        self.price_value = price_value

    def display_role(self):
        print(f"Role: Supplier from {self.company_name}")

    def supply_product(self, product):
        self.supplied_products.append(product)
        print(f"{self.company_name} supplied {product.get_name()} at ${self.price_value}")


# ---------------------------- Staff ----------------------------
class Staff(User):
    def __init__(self, id, name, email, phone, age, address, role, salary):
        super().__init__(id, name, email, phone, age, address)
        self.role = role
        self.salary = salary

    def display_role(self):
        print(f"Role: Staff - {self.role}")

    def sell_product(self, product):
        print(f"{self.get_name()} sold {product.get_name()}")


# ---------------------------- Investor ----------------------------
class Investor(User):
    def __init__(self, id, name, email, phone, age, address, amount, percentage):
        super().__init__(id, name, email, phone, age, address)
        self.invested_amount = amount
        self.invested_percentage = percentage

    def display_role(self):
        print(f"Role: Investor with ${self.invested_amount} - {self.invested_percentage}")


# ---------------------------- main ----------------------------
def main():
    # creating objects
    owner = Owner(1, "Rahul", "rahul@mail.com", 12345678, 35, "Panjim, Goa", "50%")
    customer = Customer(2, "Asha", "asha@mail.com", 12345678, 35, "Panjim, Goa")
    supplier = Supplier(3, "Vikram", "vikram@mail.com", 123456789, 35, "Panjim, Goa", "ABC Supplies", 1000)
    staff = Staff(4, "Priya", "priya@mail.com", 12345678, 45, "quepem goa", "technician", 10000)
    investor = Investor(5, "Mr. Sharma", "sharma@mail.com", 500000, 20, "savordem goa", 40000, "10%")

    store = Store(101, "TechZone", "Panjim, Goa")
    product = Product(501, "Laptop", 50000, "2 years")

    running = True

    while running:
        print("\n=== ELECTRONIC STORE SYSTEM ===")
        print("1. Customer Actions (Buy Product & Write Review)")
        print("2. Owner Actions (View Equity & Manage Store)")
        print("3. Supplier Actions (Supply Product)")
        print("4. Staff Actions (Sell Product)")
        print("5. Investor Actions (View Investment Details)")
        print("6. Exit")

        choice = input("Select an option (1-6): ")

        try:
            choice = int(choice)
        except ValueError:
            print("\nInvalid option! Please try again.")
            continue

        if choice == 1:
            print("\n--- Customer Portal ---")
            customer.display_role()
            print(f"Available Product: {product.get_name()} | Price: ${product.get_price()}")

            payment = input("Enter payment method (e.g., Cash, Card, UPI): ")

            # Buy product
            customer.buy_product(product, store, payment)

            # Review product
            rating = int(input(f"Enter rating for {product.get_name()} (1-5): "))
            comment = input("Enter review comment: ")

            customer.write_review(product, rating, comment)

        elif choice == 2:
            print("\n--- Owner Portal ---")
            owner.display_role()
            owner.add_store(store)

        elif choice == 3:
            print("\n--- Supplier Portal ---")
            supplier.display_role()
            supplier.supply_product(product)

        elif choice == 4:
            print("\n--- Staff Portal ---")
            staff.display_role()
            staff.sell_product(product)

        elif choice == 5:
            print("\n--- Investor Portal ---")
            investor.display_role()

        elif choice == 6:
            print("\nExiting system... Goodbye!")
            running = False

        else:
            print("\nInvalid option! Please try again.")


if __name__ == "__main__":
    main()

File: Electronic Store c++/electronic_store.cpp
/*
 * Electronic Store - C++ version
 * Converted from the Java "Electronic Store" project.
 *
 * This mirrors the Java structure closely since C++ (unlike C) supports
 * classes, inheritance, and virtual functions directly:
 *   - User        -> abstract base class (pure virtual displayRole())
 *   - Customer, Owner, Staff, Supplier, Investor -> derived classes (public User)
 *   - Product, Store, Review, Transaction -> plain classes, same as Java
 */

#include <iostream>
#include <string>
#include <vector>
#include <ctime>

using namespace std;

/* ---------------------------- Review ---------------------------- */
class Review {
private:
    int id;
    int rating;
    string comments;
    static int counter;

public:
    Review(int rating, const string &comments) {
        this->id = counter++;
        this->rating = rating;
        this->comments = comments;
    }
};
int Review::counter = 1;

/* ---------------------------- Transaction ---------------------------- */
class Transaction {
private:
    int id;
    string date;
    double totalAmount;
    string paymentType;
    static int counter;

public:
    Transaction(const string &paymentType, double amount) {
        this->id = counter++;
        this->paymentType = paymentType;
        this->totalAmount = amount;
        time_t now = time(nullptr);
        this->date = ctime(&now);
        if (!this->date.empty() && this->date.back() == '\n') {
            this->date.pop_back();
        }
    }
};
int Transaction::counter = 1;

/* ---------------------------- Product ---------------------------- */
class Product {
private:
    int id;
    string name;
    double price;
    double discount;
    string warrantyPeriod;
    vector<Review> reviews;

public:
    Product(int id, const string &name, double price, const string &warrantyPeriod) {
        this->id = id;
        this->name = name;
        this->price = price;
        this->discount = 0;
        this->warrantyPeriod = warrantyPeriod;
    }

    // encapsulation is used for the detail of the product data
    string getName() const { return name; }
    double getPrice() const { return price - discount; }
    void setDiscount(double discount) { this->discount = discount; }

    void addReview(const Review &r) {
        reviews.push_back(r);
    }
};

/* ---------------------------- Store ---------------------------- */
class Store {
private:
    int id;
    string name;
    string location;
    vector<Product *> products;

public:
    Store(int id, const string &name, const string &location) {
        this->id = id;
        this->name = name;
        this->location = location;
    }

    string getName() const { return name; }

    void stockProduct(Product *p) {
        products.push_back(p);
        cout << p->getName() << " stocked in " << name << endl;
    }
};

/* ---------------------------- User (abstract base class) ---------------------------- */
// this is the parent class and it is an abstract class because it has a pure
// virtual method displayRole() which will be implemented by the child classes
class User {
private:
    int id;
    string name;
    string email;
    int phone;
    int age;
    string address;

public:
    User(int id, const string &name, const string &email, int phone, int age, const string &address) {
        this->id = id;
        this->name = name;
        this->email = email;
        this->phone = phone;
        this->age = age;
        this->address = address;
    }

    virtual ~User() {}

    // encapsulation is used for the detail of the users data
    int getId() const { return id; }
    string getName() const { return name; }
    void setName(const string &name) { this->name = name; }
    string getEmail() const { return email; }
    void setEmail(const string &email) { this->email = email; }

    // Abstraction
    virtual void displayRole() const = 0;
};

/* ---------------------------- Customer ---------------------------- */
class Customer : public User {
private:
    vector<Transaction> transactions;
    vector<Review> reviews;

public:
    Customer(int id, const string &name, const string &email, int phone, int age, const string &address)
        : User(id, name, email, phone, age, address) {}

    void displayRole() const override {
        cout << "Role: Customer" << endl;
    }

    void buyProduct(Product &p, Store &store, const string &paymentType) {
        Transaction t(paymentType, p.getPrice());
        transactions.push_back(t);
        cout << getName() << " bought " << p.getName() << " from " << store.getName() << endl;
    }

    void writeReview(Product &p, int rating, const string &comment) {
        Review r(rating, comment);
        reviews.push_back(r);
        p.addReview(r);
        cout << getName() << " wrote a review for " << p.getName() << endl;
    }
};

/* ---------------------------- Owner ---------------------------- */
// so here we have created a child class of the User class
class Owner : public User {
private:
    string equity;
    vector<Store *> managedStores;

public:
    Owner(int id, const string &name, const string &email, int phone, int age, const string &address, const string &equity)
        : User(id, name, email, phone, age, address), equity(equity) {}

    // Implement the pure virtual method from the User class
    void displayRole() const override {
        cout << "Role: Owner with " << equity << " equity" << endl;
    }

    // Method to add a store to the list of managed stores
    void addStore(Store &store) {
        managedStores.push_back(&store);
        cout << getName() << " now manages " << store.getName() << endl;
    }
};

/* ---------------------------- Supplier ---------------------------- */
class Supplier : public User {
private:
    string companyName;
    vector<Product *> suppliedProducts;
    int price_value;

public:
    Supplier(int id, const string &name, const string &email, int phone, int age, const string &address,
              const string &companyName, int price_value)
        : User(id, name, email, phone, age, address), companyName(companyName), price_value(price_value) {}

    void displayRole() const override {
        cout << "Role: Supplier from " << companyName << endl;
    }

    void supplyProduct(Product &p) {
        suppliedProducts.push_back(&p);
        cout << companyName << " supplied " << p.getName() << " at $" << price_value << endl;
    }
};

/* ---------------------------- Staff ---------------------------- */
class Staff : public User {
private:
    string role;
    double salary;

public:
    Staff(int id, const string &name, const string &email, int phone, int age, const string &address,
           const string &role, double salary)
        : User(id, name, email, phone, age, address), role(role), salary(salary) {}

    void displayRole() const override {
        cout << "Role: Staff - " << role << endl;
    }

    void sellProduct(Product &p) {
        cout << getName() << " sold " << p.getName() << endl;
    }
};

/* ---------------------------- Investor ---------------------------- */
class Investor : public User {
private:
    double investedAmount;
    string investedPercentage;

public:
    Investor(int id, const string &name, const string &email, int phone, int age, const string &address,
              double amount, const string &percentage)
        : User(id, name, email, phone, age, address), investedAmount(amount), investedPercentage(percentage) {}

    void displayRole() const override {
        cout << "Role: Investor with $" << investedAmount << " - " << investedPercentage << endl;
    }
};

/* ---------------------------- main ---------------------------- */
int main() {
    // creating objects
    Owner owner(1, "Rahul", "rahul@mail.com", 12345678, 35, "Panjim, Goa", "50%");
    Customer customer(2, "Asha", "asha@mail.com", 12345678, 35, "Panjim, Goa");
    Supplier supplier(3, "Vikram", "vikram@mail.com", 123456789, 35, "Panjim, Goa", "ABC Supplies", 1000);
    Staff staff(4, "Priya", "priya@mail.com", 12345678, 45, "quepem goa", "technician", 10000);
    Investor investor(5, "Mr. Sharma", "sharma@mail.com", 500000, 20, "savordem goa", 40000, "10%");

    Store store(101, "TechZone", "Panjim, Goa");
    Product product(501, "Laptop", 50000, "2 years");

    bool running = true;

    while (running) {
        cout << "\n=== ELECTRONIC STORE SYSTEM ===" << endl;
        cout << "1. Customer Actions (Buy Product & Write Review)" << endl;
        cout << "2. Owner Actions (View Equity & Manage Store)" << endl;
        cout << "3. Supplier Actions (Supply Product)" << endl;
        cout << "4. Staff Actions (Sell Product)" << endl;
        cout << "5. Investor Actions (View Investment Details)" << endl;
        cout << "6. Exit" << endl;
        cout << "Select an option (1-6): ";

        int choice;
        cin >> choice;
        cin.ignore(); // Clear newline buffer, same purpose as scanner.nextLine() in Java

        switch (choice) {
            case 1: {
                cout << "\n--- Customer Portal ---" << endl;
                customer.displayRole();
                cout << "Available Product: " << product.getName() << " | Price: $" << product.getPrice() << endl;

                cout << "Enter payment method (e.g., Cash, Card, UPI): ";
                string payment;
                getline(cin, payment);

                // Buy product
                customer.buyProduct(product, store, payment);

                // Review product
                cout << "Enter rating for " << product.getName() << " (1-5): ";
                int rating;
                cin >> rating;
                cin.ignore(); // Clear buffer

                cout << "Enter review comment: ";
                string comment;
                getline(cin, comment);

                customer.writeReview(product, rating, comment);
                break;
            }

            case 2:
                cout << "\n--- Owner Portal ---" << endl;
                owner.displayRole();
                owner.addStore(store);
                break;

            case 3:
                cout << "\n--- Supplier Portal ---" << endl;
                supplier.displayRole();
                supplier.supplyProduct(product);
                break;

            case 4:
                cout << "\n--- Staff Portal ---" << endl;
                staff.displayRole();
                staff.sellProduct(product);
                break;

            case 5:
                cout << "\n--- Investor Portal ---" << endl;
                investor.displayRole();
                break;

            case 6:
                cout << "\nExiting system... Goodbye!" << endl;
                running = false;
                break;

            default:
                cout << "\nInvalid option! Please try again." << endl;
        }
    }

    return 0;
}
